tools/webtransport/h3/webtransport_h3_server.py

```diff
@@ -0,0 +1,247 @@
+            headers = {}
+            for header, value in event.headers:
+                headers[header] = value
+
+            method = headers.get(b":method")
+            protocol = headers.get(b":protocol")
+            if method == b"CONNECT" and protocol == b"webtransport":
+                self._handshake_webtransport(event.stream_id, headers)
+            else:
+                self._send_error_response(event.stream_id, 400)
+        elif isinstance(event, WebTransportStreamDataReceived):
+            self._handler.stream_data_received(
+                stream_id=event.stream_id, data=event.data, stream_ended=event.stream_ended)
+        elif isinstance(event, DatagramReceived):
+            self._handler.datagram_received(data=event.data)
+
+    def _send_error_response(self, stream_id: int, status_code: int) -> None:
+        headers = [
+            (b"server", SERVER_NAME.encode()),
+            (b":status", str(status_code).encode())]
+        self._http.send_headers(
+            stream_id=stream_id, headers=headers, end_stream=True)
+
+    def _handshake_webtransport(self, stream_id: int, request_headers: Dict[bytes, bytes]) -> None:
+        authority = request_headers.get(b":authority")
+        path = request_headers.get(b":path")
+        if authority is None or path is None:
+            # `:authority` and `:path` must be provided.
+            self._send_error_response(stream_id, 400)
+            return
+
+        # Create a handler using `:path`.
+        try:
+            self._handler = self._create_event_handler(
+                session_id=stream_id, path=path, request_headers=request_headers)
+        except IOError:
+            self._send_error_response(stream_id, 404)
+            return
+
+        response_headers = [
+            (b"server", SERVER_NAME.encode()),
+        ]
+        self._handler.connect_received(
+            path=path, response_headers=response_headers)
+
+        status = any(
+            header[0] == b":status" for header in response_headers)
+        if not status:
+            response_headers.append((b":status", b"200"))
+        self._http.send_headers(
+            stream_id=stream_id, headers=response_headers)
+
+        self._handler.session_established()
+
+    def _create_event_handler(self, session_id: int, path: bytes, request_headers: Dict[bytes, bytes]):
+        parsed = urlparse(path.decode())
+        file_path = os.path.join(handlers_path, parsed.path.lstrip('/'))
+        callbacks = {"__file__": file_path}
+        with open(file_path) as f:
+            exec(compile(f.read(), path, "exec"), callbacks)
+        session = WebTransportSession(self, session_id, request_headers)
+        return EventHandler(session, callbacks)
+
+
+class WebTransportSession(object):
+    """
+    A WebTransport session.
+    """
+
+    def __init__(self, protocol: WebTransportH3Protocol, session_id: int, request_headers: Dict[bytes, bytes]) -> None:
```

Comment:
Compared to the HTTP server we have much less configuration information here (c.f. tools/wptserve/wptserve/request.py). In particular access to the stash seems very useful if we need to communicate between servers.

Reply:
Makes sense. I added a TODO and will try to add more configuration as needed.